Sequential navigation in edit history

swmitra opened the pull request:
This feature is something which I missed a lot! Remembering most recently used (interesting?) cursor positions and navigating there manually is really cumbersome. 

This module adds the feature of remembering  cursors/selections across full editors inside a project root and aids the user by providing 'Navigate Back'(Alt+I) and 'Navigate Fwd'(Alt+Shift+I) using the captured navigation frames.

Ping @petetnt @ficristo for review.

swmitra committed:
First commit - sequential navigation in edit history

swmitra committed:
Fix lint errors

swmitra committed:
Handle undefined history entries

swmitra committed:
Added comments and some additional edge case handling

petetnt commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,359 @@
+        captureTimer,
+        activePosNotSynched = false,
+        jumpInProgress,
+        command_JumpBack,
```
The `camelcase` + `underscore` hybrid is a bit weird, would prefer just `commandJumpBack`.

swmitra replied:
Done.

petetnt commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,359 @@
+    */
+    var jumpedPosStack = [],
+        captureTimer,
+        activePosNotSynched = false,
```
Nit: values with defaults first

petetnt replied:
Nit: `activePosNotSynced`

swmitra replied:
Done.

petetnt commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,359 @@
+    */
+    function _validateNavigationCmds() {
+        if (jumpToPosStack.length > 0) {
+            command_JumpBack.setEnabled(true);
```
Can we just do 
``` js
commandJumpBack.setEnabled(jumpToPosStack.length > 0);
commandJumpFwd.setEnabled(jumpedPosStack.length > 0);
```

swmitra replied:
Done.

petetnt commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,359 @@
+            fileEntry = FileSystem.getFileForPath(entry.file);
+
+        if (entry.inMem) {
+            var indxInWS = MainViewManager.findInWorkingSet(entry.paneId, entry.file);
```
Nit: prefer `index` over `indx`

swmitra replied:
Done.

petetnt commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,359 @@
+    * Function to check existence of a file entry
+    * @private
+    */
+    function _checkIfExist(entry) {
```
We probably should null check for entry here

swmitra replied:
It's already checked at the call location - 

```javascript
if (navFrame) {
            // We will check for the file existence now, if it doesn't exist we will jump back again
            // but discard the popped frame as invalid.
            _checkIfExist(navFrame).done(function () {
`

petetnt commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,359 @@
+                } else {
+                    deferred.reject();
+                }
+		    });
```
Nit: indentation

swmitra replied:
Done.

petetnt commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,359 @@
+    function NavigationFrame(editor, selectionObj) {
+        this.cm = editor._codeMirror;
+        this.file = editor.document.file._path;
+        this.inMem = editor.document.file.constructor.name === "InMemoryFile" ? true : false;
```
`this.inMem = editor.document.file.constructor.name === "InMemoryFile"`

is sufficent.

swmitra replied:
Done.

petetnt commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,359 @@
+    }
+    
+   /**
+    * Binds the lifecycle event listner of the editor for which this frame is captured
```
Typo: `listener`

swmitra replied:
Done 😄

petetnt commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,359 @@
+    * -> Addition/Updation of characters in the captured selection
+    */
+    NavigationFrame.prototype._createMarkers = function (ranges) {
+        var range, index, bookMark;
```
Nit: new variables on new lines

swmitra replied:
Done.

petetnt commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,359 @@
+            range = ranges[index];
+            // 'markText' has to used for a non-zero length position, if current selection is 
+            // of zero length use bookmark instead.
+            if (range.anchor.line === range.head.line && range.anchor.ch === range.head.ch) {
```
Q: can `anchor` or `head` be null?

swmitra replied:
If there is a cursor or selection, it can never be null and if there are no selections/cursor then the ranges would be an empty array.

petetnt commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,359 @@
+        }
+        
+        this.selections = [];
+        var marker, selection, index;
```
Nit: new variables on new lines

swmitra replied:
Done.

petetnt reviewed (changes requested):
Some initial code review, good stuff @swmitra!

swmitra committed:
Update code review comments

swmitra commented:
@petetnt Thanks a lot for reviewing this PR. It was really quick 👍 
I have updated the code with relevant changes pertaining to review comments. All yours now 😄

petetnt commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,360 @@
+    // The latency time to capture an explicit cursor movement as a navigation frame
+    var NAV_FRAME_CAPTURE_LATENCY = 2000;
+    
+    /*
```
Nit: missing * (/**)

petetnt commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,360 @@
+    */
+    var jumpToPosStack = [];
+    
+    /*
```
Nit: missing * (/**)

petetnt commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,360 @@
+                } else {
+                    deferred.reject();
+                }
+			});
```
Nit: indentation: Tabs -> Spaces

petetnt commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,360 @@
+        if (!this.cm) {
+            return;
+        }
+		
```
Nit: L173-L177 indendation Tabs -> Spaces

petetnt commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,360 @@
+            }
+        });
+		
+		// Iterate over CM textmarkers and collate the updated(if?) positions
```
Nit: Indentation: tabs -> spaces

petetnt commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,360 @@
+        this.selections = [];
+        var self = this;
+		
+		// Collate only the markers we used to mark selections/cursors
```
Nit: Indentation: tabs -> spaces

petetnt commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,360 @@
+			selection,
+			index;
+        
+		// Reset selections first.
```
Nit: Indentation: tabs -> spaces

petetnt reviewed (changes requested):
Few more nits, otherwise codewise starting to LGTM. Haven't had the time to test it yet.

ficristo commented:
I will not review, but I have a question.
If I understand correctly the navigation history is never emptyed but when you change project.
Doesn't it risk to consume too much memory? Usually this kind of things have a limit, like remembering only the last 100 navigation changes.

swmitra committed:
Put a max navigation frame capture limit and control overflow once reached. Fix all bugs found in unit testing.

swmitra committed:
Fix indentation issues

swmitra committed:
Change spaces to tabs for indentation

swmitra commented:
@ficristo You are right 👍 . 
I had now put a limit of 50 and control overflow beyond that as having infinite navigation frames won't help anyone. As a developer even to do 50 back navigation would be too much.

swmitra commented:
@petetnt Can you please have a look at the PR again and if possible use it. 
@MarcelGarber Can you please have a look at this PR?

swmitra commented:
@nethip Can you please have a look at this PR?

nethip commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,432 @@
+    * @private
+    * @type {Array.<Object>}
+    */
+    var jumpedPosStack = [],
```
It is better to initialize all the variables. here.

swmitra replied:
Done 👍

petetnt commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,432 @@
+   /**
+    * Command handler to navigate forward
+    */
+    function _navigateFwd() {
```
Would prefer `_navigateForward()` here

petetnt commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,432 @@
+    
+   /**
+    * Function to create CM TextMarkers for the navigated positions/selections.
+    * This logic is required to ensure that the captured navaigation positions 
```
Typo: navigation

petetnt reviewed (changes requested):
Looks mostly good to me codewise, one nit and one typo. Also there's now mixed tabs and spaces in new code. Haven't had the time to test the functionality yet but I'll try soon.

swmitra committed:
Fix mixed tabs and spaces

swmitra commented:
@petetnt Done. Fixed the last 2 nits and mixed tabs & spaces.

swmitra committed:
Convert spaces to tab

swmitra committed:
resolve mixed tabs and spaces

swmitra committed:
Initialize variables

swmitra commented:
@nethip Can you please have a look at this PR again?

nethip commented:
@swmitra I think the code part is fine. But here are few questions I had

* What will happen if the file, for which a `NavigationFrame` is present in your `JumpToStack` or `JumpedPosStack`, is changed externally  ?

* I am trying to understand why you have to use `beforeSelection` instead of selection changed? Had it been capturing after selection change, you won't have to rely on `activePosNotSynced` variable.

* I see that you have used `currentFullEditor` for switching between different files to restore a frame state. Does that mean selections captured for editors inside `Inline` editors will transform to full view when `back` or `front` command is fired?

* I see that you are relying on `cm.setBookMark`. What happens when any extension overrides the selection marks set, while selection being changed?

* Also for batch operations like `Replace All`, are you going to log all of the selection changes?

swmitra commented:
@nethip Thanks a ton for reviewing the PR 👍 

I will try to answer the questions, please let me know if you think anything in the assumptions can go wrong.

* If we categorize external changes, we can see there are broadly 2 different kind of mutations. First one is the file entry itself being changed (delete/rename) and secondly the content being changed. In first case when we actually try to navigate using the stale frame, we end up discarding the frame as the entry doesn't exist on the disk. In second case, Brackets calls _resetText() on the existing editor with modified content which removes all the CM markers. So we gracefully discard the frame.(Let me verify it once again).

* I have used CM beforeSelection to capture the selection change "origin". This property is not part of selection change event as that's a simulated event by Brackets. We need to check for the origin of the change to discard frame capture in case user has used keyboard navigation keys or cursor change has happened because of character input. We just want to capture programmatic selection change (jump to def) or explicit pointer click. 

* This part was getting really tricky to implement as showing the selection again in InlineEditor would be impossible for two reasons - First and the most prominent one is availability of "createInlineEditor" as a public API. I have seen multiple usage of this API even by extensions and then might show this in a different DOM container all together. That's the reason I listen only for full editors ( check for _paneId). So any navigation in inline editors are not captured at all.  

* Once a textmark is created in CM, it can't be overridden. It can be removed though. In case it is removed , we won't be able to find it ("find()" will return empty position) and hence we will not act on it. 

* Correct me if I am wrong, to highlight the matches in search/replace operation, we use CM text marker instead of actual selection and to actually replace the searched token we use replaceRange. So we don't use selection in these operations. 

Please let me know if you think we can go wrong in any of these or other situations. Thanks again for raising this questions, I could verify lot of the scenarios after looking at this 😄

swmitra committed:
Handle external file changes and discard stale frames

swmitra commented:
@nethip I have added additional handling for external changes. Also, deciding on which pane the document should open on navigation is determined based on the current master editors association with pane. 

Please have a look at the additional changes.

MarcelGerber commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,553 @@
+    function _navigateForward() {
+        var navFrame = jumpedPosStack.pop();
+        
+        // Check if the poped frame is the current active frame or doesn't have any valid marker information
```
We should have an early `return` here for the case when navFrame is null (`!navFrame`)

swmitra replied:
Done.

MarcelGerber commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,553 @@
+    */
+    function _hasNavBackFrames() {
+        return (jumpedPosStack.length > 0 && jumpToPosStack.length > 0)
+            || (!jumpedPosStack.length && jumpToPosStack.length > 1);
```
I don't quite get this code.
So, when you can't jump forward, there need to be at least two elements in `jumpToPosStack`? Why is that?

MarcelGerber commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,553 @@
+    * -> Addition/Deletion of lines before the captured position
+    * -> Addition/Updation of characters in the captured selection
+    */
+    NavigationFrame.prototype._createMarkers = function (ranges) {
```
createMarkers and reinstateMarkers share a lot of code. Maybe we can consolidate them a little more?

As I see it, this should work:
``` js
NavigationFrame.prototype._reinstateMarkers = function (editor) {
    this.cm = editor._codeMirror;
    this.paneId = editor._paneId;
    this._createMarkers(this.selections)
}
```

swmitra replied:
The problem was with range start and end placeholder names but handled it now to make use of same function.

MarcelGerber commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,553 @@
+        }
+        
+        // Ensure cursor activity has not happened because of arrow keys or edit
+        if (selectionObj.origin !== "+move" && (!window.event || (window.event && window.event.type !== "input"))) {
```
You can simplify the logic here: `!window.event || window.event.type !== "input"`

swmitra replied:
Refactored this code earlier but missed this condition completely. Changed it now.

MarcelGerber commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,553 @@
+                // Check if we have reached MAX_NAV_FRAMES_COUNT
+                // If yes, control overflow
+                if (jumpToPosStack.length === MAX_NAV_FRAMES_COUNT) {
+                    var navFrame = jumpToPosStack.splice(0, 1)[0];
```
You can use `.shift()` here.

swmitra replied:
Changed.

MarcelGerber commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,553 @@
+        if (!jumpedPosStack.length) {
+            if (activePosNotSynced) {
+                currentEditPos = new NavigationFrame(EditorManager.getCurrentFullEditor(), {ranges: EditorManager.getCurrentFullEditor()._codeMirror.listSelections()});
+                jumpedPosStack.push(currentEditPos);
```
Suggestion:
Make the anonymous function inside `setTimeout` a "real" function and use it here, too.

swmitra replied:
I had to use the anonymous function to create closure in the call back scope.

MarcelGerber commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,553 @@
+    * @private
+    * @type {Array.<Object>}
+    */
+    var jumpToPosStack = [];
```
To make naming clearer, you might also use `jumpStackForward` and `jumpStackBackward`

swmitra replied:
Done!

MarcelGerber commented on src/extensions/default/NavigationAndHistory/NavigationProvider.js:
```diff
@@ -0,0 +1,553 @@
+    */
+    function NavigationFrame(editor, selectionObj) {
+        this.cm = editor._codeMirror;
+        this.file = editor.document.file._path;
```
Might call this `filepath` or just `path` to make it clear that this is _not_ a file handle.

swmitra replied:
Done. "filePath" would be more logical.

marcelgerber reviewed (changes requested):
Didn't really get to actually testing it yet.
Just one note, and I hope you are aware of it: CodeMirror already has a kind of cursor/selection history using `Ctrl-U / Ctrl-Shift-U` (Windows), so maybe we can utilize that a little? Maybe they already take care of all the markers internally?

swmitra committed:
Additional checks to validate the navigation frames

swmitra committed:
Address review comments

swmitra commented:
@MarcelGerber Thanks a lot for reviewing this PR 👍 
I have tried CM's implicit way to handle navigation but the problem was to handle navigation between multiple instances of CM which is the case we want to support predominantly. Then if we add handling of external changes and re initiating the edit session of the same file in a different CM instance, maintaining the marker cache inside CM was impossible. I had to externalize the whole thing to handle such scenarios. 
Do let me know if you think it can be handled in a better way by supporting all the features what we are going to have with this implementation. 

I have addressed most of the review comments. Please have a look at this PR and use it if possible.

schroef commented:
Just downloaded the revision, but alt+i shows me this character ˆ??

petetnt commented:
Hi @schroef,

the OSX shortcuts were improved / fixed in https://github.com/adobe/brackets/pull/13527, which didn't land to 1.10. You can change your shortcuts to match the ones in 13527 from Debug -> Show User Key Map and adding the following:

```
{
    "documentation": "https://github.com/adobe/brackets/wiki/User-Key-Bindings",
    "overrides": {
        "Cmd-Alt-Left": "navigation.jump.back",
        "Cmd-Alt-Right": "navigation.jump.fwd".
    }
}
```

schroef commented:
Sorry but that doesnt do anything as well. Looking at that code, shouldn't that also be visible then in the navigation menu, if so it's not there in 1.10

<img width="279" alt="screen shot 2017-07-08 at 9 39 11 pm" src="https://user-images.githubusercontent.com/6923008/27990472-fe3d3652-6425-11e7-9c2f-fd23742d9778.png">

schroef commented:
Seems shortcuts work not normally like in other applications. I need to pres cmd+alt+ left or right in a sequence in order to let it do something. I seem to be able to jump through documents. It doesnt always jump back to the complete history of it locations though

swmitra commented:
@schroef Brackets have a slightly different implementation for navigation. We don't remember all the positions rather only explicit positions. For example when you do mouse click at a position and pause for a brief moment to read something, do a jump to def or switch file to work on some other context.

swmitra commented:
We are going to change the OSX shorcuts and provide a small UI place holder as well in the next pre release ( within a week's time!)

schroef commented:
Well it doesnt store all the position i think. I've tried it couple times and sometimes is seems to work and sometimes not. I definitely dont see that history of 30 states.

utkarshmehta reviewed (commented):
a
